articles/react-error-handling.md

```diff
@@ -0,0 +1,436 @@
+
+We can add a catch method on the promise execution and have two new states to record the error message and the status.
+
+```tsx
+function App() {
+  const [productId, setProductId] = useState('')
+  const [productData, setProductData] = React.useState<ProductData | null>(null)
+  const [error, setError] = useState<string | null>(null)
+  const [status, setStatus] = useState('idle')
+  const inputRef = React.useRef<HTMLInputElement>(null)
+
+  const handleSearch = () => {
+    setProductId(inputRef.current?.value ?? '')
+  }
+
+  React.useEffect(() => {
+    if (!productId) {
+      return
+    }
+
+    setStatus('pending')
+    makeRequest(productId)
+      .then((data) => {
+        setProductData(data)
+        setStatus('resolved')
+      })
+      .catch((error: Error) => {
+        setError(error.message)
+        setStatus('rejected')
+      })
+  }, [productId])
+
+  const renderContent = () => {
+    console.log({error, status, productData})
+    switch (status) {
+      case 'pending':
+        return (
+          <p>Loading</p>
+        )
+      case 'resolved':
+        return (
+          <ProductInfo product={productData!} />
+        )
+      case 'rejected':
+        return (
+          <div>{error}</div>
+        )
+      default:
+        return (
+          <p>Type a product ID</p>
+        )
+    }
+  }
+
+  return (
+    <div>
+      <div>
+        <input type="text" ref={inputRef} />
+        <button onClick={handleSearch}>Search</button>
+      </div>
+      <div>
+        {renderContent()}
+      </div>
+    </div>
+  )
+}
+```
+
+![product-search-whole-flow.gif](/images/articles/react-error-handling/product-search-whole-flow.gif)
+
+
+It works on a way better organized :). 
+
+Getting to the point of this whole post what if the backend team makes some release on the API and some productReference's value is no longer being returned (if you are a backend developer and you are reading this, don't hesitate this is just for testing purpose you are the best of the best). This affects our new functionallity to show some product details based on the productReference. Let's take a look how this behaves, we will not actually change anyything but removing a productReference's value.
+
+```ts
+const products = {
+  '1': {
+    productName: 'xxxx',
+    productReference: '1-xxxx',
+    price: 0
+  },
+  '2': {
+    productName: 'xxxx',
+    price: 0
+  }
+} as unknown as Record<string, ProductData>
+```
+The casting on the last line is to not update any ts interface. If we review our frontend result, we get the following:
+
+![product-search-uncaught-product-reference-error.gif](/images/articles/react-error-handling/product-search-uncaught-product-reference-error.gif)
+
+May it be our error? Yes, it may. Must we create a ticket to the backend team? As soon as possible please. While the error gets solved let's review the error handling pattern installing the next [dependecy](https://www.npmjs.com/package/react-error-boundary).
```

Comment:
```suggestion
Could it be our mistake? Yes, it may. Must we create a ticket to the backend team? As soon as possible please. While the error gets solved let's review the error handling pattern installing the next [dependecy](https://www.npmjs.com/package/react-error-boundary).
```

Reply:
```suggestion
Could it be our mistake? Yes, it could. Must we create a ticket to the backend team? As soon as possible please. While the error gets solved let's review the error handling pattern installing the next [dependency](https://www.npmjs.com/package/react-error-boundary).
```